File: analysis/dependency_graph.py
import networkx as nx
from typing import Any
from datetime import datetime


def build_dependency_graph(requests: list[dict], storage_events: list[dict] | None = None) -> nx.DiGraph:
    G = nx.DiGraph()

    for req in requests:
        url = req.get("url") or req.get("payload", {}).get("url", "")
        if not url:
            continue
        resource_type = req.get("resourceType") or req.get("payload", {}).get("resourceType", "unknown")
        G.add_node(url, type=resource_type, ts=req.get("ts"))

        initiator_url = req.get("initiator", {}).get("url") if isinstance(req.get("initiator"), dict) else None
        if initiator_url:
            G.add_edge(initiator_url, url, type="initiates", ts=req.get("ts"))

    if storage_events:
        for ev in storage_events:
            payload = ev.get("payload", {})
            key = payload.get("key", "unknown")
            storage_type = payload.get("storage", "localStorage")
            script_url = ev.get("url")
            if script_url:
                G.add_edge(script_url, f"storage:{storage_type}:{key}", type="writes")

    return G


def analyze_dependency_graph(G: nx.DiGraph) -> dict[str, Any]:
    if G.number_of_nodes() == 0:
        return {"nodes": 0, "edges": 0}

    return {
        "nodes": G.number_of_nodes(),
        "edges": G.number_of_edges(),
        "density": nx.density(G),
        "roots": [n for n in G.nodes if G.in_degree(n) == 0],
        "leaves": [n for n in G.nodes if G.out_degree(n) == 0],
        "third_parties": [n for n in G.nodes if is_third_party(n)],
        "components": nx.number_weakly_connected_components(G),
    }


def is_third_party(url: str) -> bool:
    import re
    third_party_domains = [
        "google-analytics.com", "gtag", "segment.io", "mixpanel.com",
        "doubleclick.net", "googlesyndication.com", "amazon-adsystem.com",
        "cloudflare.com", "fastly.net", "akamaihd.net",
        "facebook.net", "hotjar.com",
        "intercom.io", "drift.com", "crisp.chat",
    ]
    return any(d in url for d in third_party_domains)

File: api-rust/src/analysis/dependency_graph.rs
use petgraph::graph::{DiGraph, NodeIndex};
use petgraph::algo;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DependencyGraphResult {
    pub nodes: usize,
    pub edges: usize,
    pub density: f64,
    pub roots: Vec<String>,
    pub leaves: Vec<String>,
    pub third_parties: Vec<String>,
    pub components: usize,
}

fn is_third_party(url: &str) -> bool {
    let known_domains = [
        "google-analytics.com", "gtag", "segment.io", "mixpanel.com",
        "doubleclick.net", "googlesyndication.com", "amazon-adsystem.com",
        "cloudflare.com", "fastly.net", "akamaihd.net", "facebook.net",
        "hotjar.com", "intercom.io", "drift.com", "crisp.chat",
    ];
    known_domains.iter().any(|d| url.contains(d))
}

pub fn build_and_analyze(
    requests: &[serde_json::Value],
    storage_events: Option<&[serde_json::Value]>,
) -> DependencyGraphResult {
    let mut graph = DiGraph::<String, String>::new();
    let mut node_indices: HashMap<String, NodeIndex> = HashMap::new();

    for req in requests {
        let url = req.get("url").and_then(|v| v.as_str()).unwrap_or("").to_string();
        if url.is_empty() {
            continue;
        }

        let idx = *node_indices.entry(url.clone()).or_insert_with(|| graph.add_node(url.clone()));

        let resource_type = req.get("resourceType").and_then(|v| v.as_str()).unwrap_or("");
        if let Some(node) = graph.node_weight_mut(idx) {
            if !node.contains(resource_type) {
                node.push_str(&format!(" [{}]", resource_type));
            }
        }

        // Edge: initiator_url -> url
        if let Some(initiator) = req
            .get("initiator")
            .or_else(|| req.pointer("/payload/initiator"))
            .and_then(|i| i.get("url"))
            .and_then(|v| v.as_str())
        {
            if !initiator.is_empty() {
                let init_idx = *node_indices
                    .entry(initiator.to_string())
                    .or_insert_with(|| graph.add_node(initiator.to_string()));
                if !graph.contains_edge(init_idx, idx) {
                    graph.add_edge(init_idx, idx, "initiates".to_string());
                }
            }
        }
    }

    // Storage edges
    if let Some(storage) = storage_events {
        for ev in storage {
            let script_url = ev
                .get("payload")
                .and_then(|p| p.get("scriptUrl"))
                .and_then(|v| v.as_str())
                .unwrap_or("");

            let storage_type = ev
                .get("payload")
                .and_then(|p| p.get("storageType"))
                .and_then(|v| v.as_str())
                .unwrap_or("unknown");

            let key = ev
                .get("payload")
                .and_then(|p| p.get("key"))
                .and_then(|v| v.as_str())
                .unwrap_or("unknown");

            if !script_url.is_empty() {
                let storage_node_id = format!("storage:{}:{}", storage_type, key);
                let src_idx = *node_indices
                    .entry(script_url.to_string())
                    .or_insert_with(|| graph.add_node(script_url.to_string()));
                let dst_idx = *node_indices
                    .entry(storage_node_id.clone())
                    .or_insert_with(|| graph.add_node(storage_node_id));
                if !graph.contains_edge(src_idx, dst_idx) {
                    graph.add_edge(src_idx, dst_idx, "writes".to_string());
                }
            }
        }
    }

    let nodes = graph.node_count();
    let edges = graph.edge_count();

    if nodes == 0 {
        return DependencyGraphResult {
            nodes: 0, edges: 0, density: 0.0,
            roots: vec![], leaves: vec![], third_parties: vec![], components: 0,
        };
    }

    let density = if nodes > 1 {
        let max_edges = nodes * (nodes - 1);
        edges as f64 / max_edges as f64
    } else {
        0.0
    };

    let roots: Vec<String> = graph
        .node_indices()
        .filter(|&n| graph.neighbors_directed(n, petgraph::Direction::Incoming).count() == 0)
        .filter_map(|n| graph.node_weight(n).cloned())
        .collect();

    let leaves: Vec<String> = graph
        .node_indices()
        .filter(|&n| graph.neighbors_directed(n, petgraph::Direction::Outgoing).count() == 0)
        .filter_map(|n| graph.node_weight(n).cloned())
        .collect();

    let third_parties: Vec<String> = graph
        .node_indices()
        .filter_map(|n| graph.node_weight(n))
        .filter(|n| is_third_party(n))
        .cloned()
        .collect();

    let components = petgraph::algo::connected_components(&graph);

    DependencyGraphResult {
        nodes,
        edges,
        density: (density * 1000.0).round() / 1000.0,
        roots,
        leaves,
        third_parties,
        components,
    }
}
